### DISCUSSION TITLE
How to get the point on G closest to H?

### DISCUSSION THREAD

**Discussion by jpmaterial** [shapely/shapely#1287] [Q&A]
I don't just want to know the distance given by `G.distance(H)` but also the two points in G and H between which the distance is actually measured.

I tried this:

    l0 = LineString([(0,0),(2,0)]
    l1 = LineString([(1,1),(1,3)]
    p0 = l0.interpolate( l0.project( l1 ) )
    p1 = l1.interpolate( l1.project( l0 ) )

Here `p0` will actually be what I'm looking for (`(1,0)`), but `p1` won't (but `(1,2)`).

**Comment by jpmaterial** [comment]
Okay, for LineString this seems to work:

    def point_closest_to(this, other):

        d = this.distance( other )

        if d == 0:
            return Point(this.intersection(other).coords[0])

        for coord in this.coords:
            end_point = Point(coord)
            ints = end_point.buffer( d ).exterior.intersection( other )
            if type( ints ) == Point:
                return end_point

        for coord in other.coords:
            end_point = Point(coord)
            ints = end_point.buffer( d ).exterior.intersection( this )
            if type( ints ) == Point:
                return ints

        raise Exception('Not found')

Does this look correct? Is there a simpler way?

**Comment by jorisvandenbossche** [comment]
You can use the `shapely.ops.nearest_points` function for this:

```python
>>> shapely.ops.nearest_points(l0, l1)
(<shapely.geometry.Point POINT (1 0)>, <shapely.geometry.Point POINT (1 1)>)
```

Starting with Shapely 2.0 (upcoming release), there will also be a `shortest_line` function that gives the same two points but as a line:

```python
>>> shapely.shortest_line(l0, l1)
<shapely.geometry.LineString LINESTRING (1 0, 1 1)>
```
